lxd/firewall/interfaces.go

```diff
@@ -0,0 +1,32 @@
+package firewall
+
+import (
+	"net"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+)
+
+
+// Firewall represents an LXD firewall.
+type Firewall interface {
+	// Lower-level Functions
+	NetworkClear(protocol string, comment string, table string) error
+	ContainerClear(protocol string, comment string, table string) error
+	VerifyIPv6Module() error
+
+	// Proxy Functions
+	InstanceProxySetupNAT(protocol string, ipAddr net.IP, comment string, connType, address, port string, cPort string) error
+
+	// NIC Bridged Functions
+	InstanceNicBridgedRemoveFilters(m deviceConfig.Device, ipv4 net.IP, ipv6 net.IP) error
+	InstanceNicBridgedSetFilters(m deviceConfig.Device, config map[string]string, ipv4 net.IP, ipv6 net.IP, name string) error
+
+	// Network FunctionsFunctions
+	NetworkSetupConfigIPv4Firewall(name string, config map[string]string) error
+	NetworkSetupAllowIPv4Forwarding(name string, config map[string]string) error
+	NetworkSetupConfigIPv4NAT(name string, config map[string]string, subnet net.IPNet) error
+	NetworkSetupConfigIPv6Firewall(name string) error
+	NetworkSetupAllowIPv6Forwarding(name string, config map[string]string) error
+	NetworkSetupConfigIPv6NAT(name string, config map[string]string, subnet net.IPNet) error
+	NetworkSetupConfigTunnelNAT(name string, config map[string]string, overlaySubnet net.IPNet) error
```

Comment:
For those, I don't like the idea of exposing the config map outside of `networks.go` as that would result in us having to put LXD networking logic directly into the firewall code (things like bridge mode, fan, ...).

I'd instead be looking at functions that the `networks.go` code could call to do what it actually wants to do.
So thing like adding an ALLOW rule in the INPUT table for some protocol and port, setting up nat for some subnet, using an optional source address, ...

Reply:
Do you think we should just collapse the entire NetworkSetup into one abstracted away function? Our initial concern with that idea is that it does some general things between xtables and nft that would end up being copied code from one to the other. We identified the blocks of code specific to the xtables implementation and pulled them out in logical groupings.

Reply:
I think those likely need to be split into more functions so that you never have to pass the network config map to the firewall package.

Reply:
This change has been made in a recent commit